torch/lib/THD/base/channels/DataChannelTCP.cpp

```diff
@@ -196,67 +196,99 @@ DataChannelTCP::~DataChannelTCP()
 
 
 void DataChannelTCP::listen(std::uint16_t port = 0) {
-  SYSCHECK(_socket = ::socket(PF_INET, SOCK_STREAM, 0))
+  struct addrinfo hints, *res = NULL;
+
+  memset(&hints, 0x00, sizeof(hints));
+  hints.ai_flags = AI_PASSIVE;
+  hints.ai_family = AF_UNSPEC; // either IPv4 or IPv6
+  hints.ai_socktype = SOCK_STREAM; // TCP
+
+  // `getaddrinfo` will sort addresses according to RFC 3484 and can be tweeked
+  // by editing `/etc/gai.conf`. so there is no need to manual sorting
+  // or protcol preference.
+  int err = getaddrinfo(NULL, std::to_string(port).data(), &hints, &res);
+  if (err != 0 || !res) {
```

Comment:
It could but it would involve some unnecessary code since we want to print `gai_strerror`. Because to do that we would have to do:
```
int err;
try {
  SYSCHECK(...)
} catch (system_error) {
  invalid_argument(... gai_strerror)
}

if (!res) {
  invalid_argument(...)
}
```